cellrank/tl/kernels/_base_kernel.py

```diff
@@ -294,21 +295,37 @@ def compute_projection(
                 "Compute transition matrix first as `.compute_transition_matrix()`."
             )
 
+        if self.kernels[0]._conn is None:
+            raise AttributeError(
+                f"Connectivity matrix not defined for {self.kernels[0]}. As of now, "
+                "the embedding projection only works for kNN based kernels."
+            )
+
         start = logg.info(f"Projecting transition matrix onto `{basis}`")
         emb = _get_basis(self.adata, basis)
         T_emb = np.empty_like(emb)
 
+        T = self.transition_matrix.copy()
+        T.sort_indices()
+
+        conn = self.kernels[0]._conn.copy()
+        conn.sort_indices()
+
         with warnings.catch_warnings():
             warnings.simplefilter("ignore")
-            for i, row in enumerate(self.transition_matrix):
-                dX = emb[row.indices] - emb[i, None]
+            for row_id, row in enumerate(T):
+                conn_idxs = conn[row_id, :].indices
+                dX = emb[conn_idxs] - emb[row_id, None]
+
+                probs = np.zeros_like(conn_idxs, dtype=T.dtype)
+                probs[np.isin(conn_idxs, row.indices)] = row.data
```

Comment:
This assumes that `len(row.data) == np.sum(np.isin(conn_idxs, row.indices)])`, which imho needs not to hold.
Think `probs = row[:, conn_idxs].data` should do the same + might make it easier when not sorting indices/making a copy.

Reply:
Hm, yes I had only kNN based kernels in mind. There it should be okay, I think. We can do

```python
probs = np.zeros_like(conn_idxs, dtype=row.dtype)
overlap = np.isin(conn_idxs, row.indices)
probs[overlap] = row[:, conn_idxs[overlap]].data
```

but still need to sort, though, to make sure indices in `_conn` and `row` are ordered the same. `row[:, conn_idxs].data` subsets to the right data but returns them w.r.t. the indices of `row`.

Also

```python
probs = row[:, conn_idxs].data
```

would again only consider non-trivial transition probabilities, *i.e.*, the shape of `dX` and `probs` would mismatch for the `PseudoTimeKernel`.

Reply:
Your right, my mistake, should be `row[:, conn_idxs].A.squeeze()`. Here's my proposal (should be identical):
```python
            for row_id, row in enumerate(self.transition_matrix):
                conn_idxs = conn[row_id, :].indices
                dX = emb[conn_idxs] - emb[row_id, None]

                if np.any(np.isnan(dX)):
                    T_emb[row_id, :] = np.nan
                else:
                    probs = row[:, conn_idxs]
                    if issparse(probs):
                        probs = probs.A.squeeze()
                    dX /= np.linalg.norm(dX, axis=1)[:, None]
                    dX = np.nan_to_num(dX)
                    T_emb[row_id, :] = probs.dot(dX) - dX.sum(0) / dX.shape[0]
```
One problem I have with `row.sort_indices(); overlap = np.isin(conn_idxs, row.indices)` is that it relies on `.transition_matrix` (was done in https://github.com/theislab/cellrank/pull/586, but wasn't fixed for emb. projection), altough 99% of the time it will be sparse (based on types, not necessarily CSR).

Reply:
Good point. I refactored the code as you suggested.